### PR TITLE
[monodroid] Support for .NET6

### PR DESCRIPTION
**Pull request by grendello**
* Use Mono headers and libs from .NET6 runtime packages
 * Update CMake to 3.18.1
 * Refactor `CMakeLists.txt` to take advantage of new CMake features
   and clean up the scripts.
 * Use `xaprepare` to generate `CMake` commands (msbuild and shell
   script)
   * The entire msbuild `_PrepareConfigureRuntimeCommands` which used to
     be in `monodroid.targets` is now generated
   * A new `build-tools/scripts/build-monodroid` script added, enabling
     one to build any `libmonodroid` configuration from command line,
     without the need to use `msbuild` (which builds all the runtimes,
     currently 52 configurations).
   * Both the `msbuild` and shell bits are generated from the same
     source, making maintenance easier.

The runtime is already built against .NET6 libraries and headers, no .NET6 specific code in the runtime yet.

TODO:
 - [x] Package the .NET6 runtimes into nugets
 - [ ] Add .NET6 specific code to `src/monodroid`
   - [x] New VM initialization calls
   - [ ] Support for the new interpreter modes
   - [x] `PINVOKE_OVERRIDE` support
   - [ ] Add pinvoke override lookup caching
   - [ ] Use AssemblyLoadContext versions of image/assembly loader functions

### CODE REVIEW

File: Configuration.props [hunk 1]
```diff
@@ -75,13 +75,19 @@
     <DotNetPreviewPath Condition=" '$(DotNetPreviewPath)' == '' ">$(AndroidToolchainDirectory)\dotnet\</DotNetPreviewPath>
     <DotNetPreviewTool Condition=" '$(DotNetPreviewTool)' == '' ">$(DotNetPreviewPath)dotnet</DotNetPreviewTool>
     <!-- Version number from: https://github.com/dotnet/installer#installers-and-binaries -->
+    <!-- Please update DotNetRuntimePacksVersion below accordingly -->
     <DotNetPreviewVersionBand Condition=" '$(DotNetPreviewVersionBand)' == '' ">6.0.100</DotNetPreviewVersionBand>
     <DotNetPreviewVersionFull Condition=" '$(DotNetPreviewVersionFull)' == '' ">$(DotNetPreviewVersionBand)-preview.1.21103.13</DotNetPreviewVersionFull>
+    <!-- This version comes from the a file in the dotnet distribution: sdk/$(DotNetPreviewVersionFull)/dotnet.runtimeconfig.json -->
+    <DotNetRuntimePacksVersion Condition=" '$(DotNetRuntimePacksVersion)' == '' ">[6.0.0-preview.2.21080.7]</DotNetRuntimePacksVersion>
```
Comment (jonathanpeppers) [inline Configuration.props:82]: I think you could do:

```suggestion
    <DotNetRuntimePacksVersion Condition=" '$(DotNetRuntimePacksVersion)' == '' ">6.0.0-preview.2.21080.7</DotNetRuntimePacksVersion>
```

If you put the `[]` in the `@(PackageDownload)`, you wouldn't have to trim these `[]` in other places. This worked for me [here](https://github.com/jonathanpeppers/maui-workload/blob/ec0530198cdf17f7e2dba89807f6282afb933922/src/Maui.Dependencies/Maui.Dependencies.Sdks.csproj#L6-L11):

```diff
<PackageDownload Include="Microsoft.NETCore.App.Runtime.android-arm64" Version="[$(DotNetRuntimePacksVersion)]" />
```

It also kind of makes sense to put the weird NuGet requirement in this one spot.

File: build-tools/xaprepare/xaprepare/xaprepare.csproj [hunk 2]
```diff
@@ -62,4 +62,10 @@
   </ItemGroup>
 
   <Import Project="xaprepare.targets" Condition=" $(MSBuildToolsPath.IndexOf('omnisharp')) &lt; 0 " />
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp3.1' ">
+    <PackageDownload Include="Microsoft.NETCore.App.Runtime.android-arm64" Version="$(DotNetRuntimePacksVersion)" />
+    <PackageDownload Include="Microsoft.NETCore.App.Runtime.android-x64" Version="$(DotNetRuntimePacksVersion)" />
+    <PackageDownload Include="Microsoft.NETCore.App.Runtime.android-arm" Version="$(DotNetRuntimePacksVersion)" />
+    <PackageDownload Include="Microsoft.NETCore.App.Runtime.android-x86" Version="$(DotNetRuntimePacksVersion)" />
```
Comment (jonathanpeppers) [inline build-tools/xaprepare/xaprepare/xaprepare.csproj:69]: If you get an error, you might need to add the `[]` to these lines.

File: build-tools/xaprepare/xaprepare/ConfigAndData/Configurables.cs [hunk 3]
```diff
@@ -386,6 +398,17 @@ public static partial class Paths
 				return Directory.Exists (path) ? path : Path.Combine (MonoSdksTpnExternalPath, "llvm");
 			});
 
+			static string GetNetcoreAppRuntimePath (Context ctx, string androidTarget)
+			{
+				return Path.Combine (
+					XAPackagesDir,
+					$"microsoft.netcore.app.runtime.android-{androidTarget}",
+					ctx.Properties.GetRequiredValue (KnownProperties.DotNetRuntimePacksVersion).Trim ('[').Trim (']'),
```
Comment (jonathanpeppers) [inline build-tools/xaprepare/xaprepare/ConfigAndData/Configurables.cs:406]: Probably don't need to `Trim()` anymore.

File: build-tools/xaprepare/xaprepare/xaprepare.csproj (hunk 2 above)
Comment (grendello) [reply build-tools/xaprepare/xaprepare/xaprepare.csproj]: Yep, just pushed the change, thanks :)

File: build-tools/create-packs/Directory.Build.targets [hunk 4]
```diff
@@ -77,10 +77,10 @@
...
       DependsOnTargets="BuildILLinkCustomStep;DeleteExtractedWorkloadPacks;_SetGlobalProperties">
     <RemoveDir Directories="$(XamarinAndroidSourcePath)bin\Build$(Configuration)\nupkgs" />
-    <Exec Command="$(DotNetPreviewTool) pack @(_GlobalProperties, ' ') -p:AndroidRID=android.21-arm -p:AndroidABI=armeabi-v7a &quot;$(MSBuildThisFileDirectory)Microsoft.Android.Runtime.proj&quot;" />
-    <Exec Command="$(DotNetPreviewTool) pack @(_GlobalProperties, ' ') -p:AndroidRID=android.21-arm64 -p:AndroidABI=arm64-v8a &quot;$(MSBuildThisFileDirectory)Microsoft.Android.Runtime.proj&quot;" />
-    <Exec Command="$(DotNetPreviewTool) pack @(_GlobalProperties, ' ') -p:AndroidRID=android.21-x86 -p:AndroidABI=x86 &quot;$(MSBuildThisFileDirectory)Microsoft.Android.Runtime.proj&quot;" />
-    <Exec Command="$(DotNetPreviewTool) pack @(_GlobalProperties, ' ') -p:AndroidRID=android.21-x64 -p:AndroidABI=x86_64 &quot;$(MSBuildThisFileDirectory)Microsoft.Android.Runtime.proj&quot;" />
+    <Exec Command="$(DotNetPreviewTool) pack @(_GlobalProperties, ' ') -p:AndroidRID=android.21-arm -p:AndroidABI=armeabi-v7a-net6 &quot;$(MSBuildThisFileDirectory)Microsoft.Android.Runtime.proj&quot;" />
```
Comment (jonpryor) [inline build-tools/create-packs/Directory.Build.targets:80]: This should drop the `.21`, as per commit 72da34928f3ed0abce5e0c0db313fbf54b428b58.

File: build-tools/create-packs/Microsoft.Android.Runtime.proj [hunk 5]
```diff
@@ -11,7 +11,7 @@ projects that use the Microsoft.Android framework in .NET 5.
 
   <PropertyGroup>
     <AndroidRID Condition=" '$(AndroidRID)' == '' ">android.21-arm64</AndroidRID>
-    <AndroidABI Condition=" '$(AndroidABI)' == '' ">arm64-v8a</AndroidABI>
+    <AndroidABI Condition=" '$(AndroidABI)' == '' ">arm64-v8a-ne6</AndroidABI>
```
Comment (jonpryor) [inline build-tools/create-packs/Microsoft.Android.Runtime.proj:14]: Should this be `-net6`, not `-ne6`?  `-net6` is consistent with the changes in `Directory.Build.props`.

File: src/monodroid/jni/monodroid-glue.cc [hunk 6]
```diff
@@ -1635,6 +1651,19 @@ MonodroidRuntime::Java_mono_android_Runtime_initInternal (JNIEnv *env, jclass kl
...
 	}
 
+#if defined (NET6)
+	{
+		PInvokeOverrideFn pinvoke_override_cb = monodroid_pinvoke_override;
+		char ptr_str[20];
+		snprintf (ptr_str, sizeof(ptr_str), "0x%p", pinvoke_override_cb);
+
+		const char* property_keys[] { "PINVOKE_OVERRIDE" };
+		const char* property_values[] { ptr_str };
```
Comment (jonpryor) [inline src/monodroid/jni/monodroid-glue.cc:1661]: this remains gloriously insane.

Comment (grendello) [reply src/monodroid/jni/monodroid-glue.cc:1661]: It is... Thus, https://github.com/dotnet/runtime/issues/48416 :)

**Pull request closed without merging**